python-lib/osisoft_client.py

```diff
@@ -43,6 +43,144 @@ def get_auth(self, auth_type, username, password):
+                    logger.info("Yielding first item")
+                    yield first_item
+                counter += 1
+                for first_item in ret:
+                    yield first_item
+                    counter += 1
+            except Exception as err:
+                if "Error 400" in "{}".format(err):
+                    logger.warning("The time range {} -> {} is too large, splitting the job in two".format(start_date, end_date))
+                    start_timestamp = self.parse_pi_time(start_date)
+                    end_timestamp = self.parse_pi_time(end_date)
+                    new_time_range = (end_timestamp - start_timestamp) / 2
+                    half_time_iso = epoch_to_iso(start_timestamp + new_time_range)
+                    first_half_items = self.get_all_rows_from_webid(
+                        webid, data_type, start_date=epoch_to_iso(start_timestamp), end_date=half_time_iso,
+                        interval=interval, sync_time=sync_time, boundary_type=boundary_type, selected_fields=selected_fields,
+                        can_raise=can_raise, endpoint_type=endpoint_type, search_full_hierarchy=search_full_hierarchy, max_count=max_count
+                    )
+                    for item in first_half_items:
+                        yield item
+                    logger.info("Successfuly retrieved first half of {} to {}".format(epoch_to_iso(start_timestamp), half_time_iso))
+                    second_half_items = self.get_all_rows_from_webid(
+                        webid, data_type, start_date=half_time_iso, end_date=epoch_to_iso(end_timestamp),
+                        interval=interval, sync_time=sync_time, boundary_type=boundary_type, selected_fields=selected_fields,
+                        can_raise=can_raise, endpoint_type=endpoint_type, search_full_hierarchy=search_full_hierarchy, max_count=max_count
+                    )
+                    for item in second_half_items:
+                        yield item
+                    logger.info("Successfuly retrieved second half of {} to {}".format(half_time_iso, epoch_to_iso(end_timestamp)))
+                else:
+                    logger.error("Error: {}".format(err))
+                    raise Exception("Error: {}".format(err))
+            logger.info("Successfuly retrieved time range {} to {}".format(start_date, end_date))
+            if counter == max_count:
+                logger.warning("Number of replies equals maxCount. Shifting startDate and trying one more time.")
+                last_received_timestamp = first_item.get("Timestamp")
+                logger.info("Last received timestamp is {}".format(last_received_timestamp))
+                start_date = last_received_timestamp
+                previous_item_timestamp = last_received_timestamp
+            else:
+                not_happy = False
+
+    def get_all_rows_from_item(self, pi_tag, data_type, start_date=None, end_date=None,
```

Comment:
Here `item` can be a tag, or a path to an attribute, an element or an event frame

Reply:
Thanks - I see. Then we should probably change the name of the argument `pi_tag` to `item` and add comments to explain the range of things it could be (the latter also applies to `get_row_from_item`).

Reply:
👍 done in [ef7ecbc](https://github.com/dataiku/dss-plugin-osisoft/pull/43/commits/ef7ecbc57adf80d85be841e51174a3396fb1bba0)